File: 700/726 - Number of Atoms.py
class Solution:
    def _get_count(self, formula, i):
        t = i
        while i < len(formula) and formula[i].isnumeric():
            i += 1
        count = 1 if t == i else int(formula[t:i])

        return i, count
    
    def _get_element(self, formula, i):
        name = formula[i]
        i += 1
        if i < len(formula) and formula[i].islower():
            name += formula[i]
            i += 1

        i, count = self._get_count(formula, i)

        return i, name, count
    
    def parse(self, formula, i):
        ret = {}

        while i < len(formula):

            if formula[i] == '(':
                i += 1
                i, tmp = self.parse(formula, i)

                for name, count in tmp.items():
                    ret[name] = ret.get(name, 0) + count

            elif formula[i] == ')':
                i += 1
                i, count = self._get_count(formula, i)
                for name in ret:
                    ret[name] *= count

                return i, ret

            else:
                i, name, count = self._get_element(formula, i)
                ret[name] = ret.get(name, 0) + count

        return i, ret
    
    def countOfAtoms(self, formula):
        """
        :type formula: str
        :rtype: str
        """
        _, ret = self.parse(formula, 0)

        return "".join(sorted(name if count == 1 else "%s%d" % (name, count) for name, count in ret.items()))

File: csharp/726.number-of-atoms.cs
/*
 * @lc app=leetcode id=726 lang=csharp
 *
 * [726] Number of Atoms
 */

using System.Collections.Generic;
using System.Linq;
using System;

// @lc code=start
public class Solution {

    private (IDictionary<string, int> atoms, int index) CountOfAtomsRecurse(string formula, int i) {
        IDictionary<string, int> ret = new Dictionary<string, int>();
        int n = formula.Length;
        while (i < n) {
            char c = formula[i];
            if (c == '(') {
                var inner = CountOfAtomsRecurse(formula, i + 1);
                foreach (var tuple in inner.atoms) {
                    ret.TryGetValue(tuple.Key, out int count);
                    count += tuple.Value;
                    ret[tuple.Key] = count;
                }
                i = inner.index;
            } else if (c == ')') {
                ++i;
                int j = i;
                while (j < n && formula[j] >= '0' && formula[j] <= '9') {
                    ++j;
                }
                if (j > i) {
                    int mul = int.Parse(formula.Substring(i, j - i));
                    foreach (var name in ret.Keys.ToArray()) {
                        ret[name] *= mul;
                    }
                    i = j;
                }
                break;
            } else {
                int j = i + 1;
                if (j < n && formula[j] >= 'a' && formula[j] <= 'z') {
                    ++j;
                }
                string name = formula.Substring(i, j - i);
                ret.TryGetValue(name, out int count);
                i = j;
                while (j < n && formula[j] >= '0' && formula[j] <= '9') {
                    ++j;
                }
                count += i == j ? 1 : int.Parse(formula.Substring(i, j - i));
                ret[name] = count;
                i = j;
            }
        }
        return (atoms: ret, index: i);
    }

    public string CountOfAtoms(string formula) {
        return string.Concat(CountOfAtomsRecurse(formula, 0).atoms.Select(tuple => tuple.Value > 1 ? $"{tuple.Key}{tuple.Value}" : tuple.Key).OrderBy(s => s, StringComparer.Ordinal));
    }
}
// @lc code=end

